### ISSUE TITLE
SAB Question which relates to Billing

### ISSUE THREAD

**Issue by kdjstudios** [kdjstudios/SABillingGitlab#757]
In GitLab by @kdjstudios on Mar 15, 2018, 10:52

**Submitted by:**  "Kimberly Gagner" <kimberly.gagner@answernet.com>

**Helpdesk:** http://www.servicedesk.answernet.com/profiles/ticket/2018-03-15-59818/conversation

**Server:** Internal

**Client/Site:** 123

**Account:** 

**Issue:** 

I have a client who had questioned their usage on their February 1, 2018 invoice.  I ran a detailed client summary report in Pinnacle Reports for the billing period and the number does not match what the billable time was charged in SAB. 

 

For example, the report you add the In and Out time you come up with 975.8 which I round up to 976.  However, if you look at the invoice they are billed for 1009 units which is a difference of 33 units.  I looked at the report to see if a column had 33 in it and maybe you are billing for something in the report I am unaware of and I don’t see anything for 33. 

 

I need to know the exact process what information is uploaded into SAB what mathematical calculations are used?  Do you round up or down on any line items?  Do you add a % so you do not have to round up or down on each line item?  Once I know this information, I can present it to Professional Teledata and they will be able to set up a report to match the invoice that goes to the client.

 

Please let me know if you have any additional questions,

[invoice_123-67311_Feb.pdf](/uploads/d19d687e8317e10fa75a2c52a5764ce1/invoice_123-67311_Feb.pdf)

[Copy+of+billing_detail__summary_lynnhousing+_28002_29.xls](/uploads/e86f86eef12fa8cbedf61a8e85f46420/Copy+of+billing_detail__summary_lynnhousing+_28002_29.xls)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 15, 2018, 10:53

@smasih @tim.traylor 

I tried to look into the account's Human Readable but all of the calculated values show Zero, however it is showing usage on those calculated fields in the invoice..?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 15, 2018, 10:57

My initial response to Kimberly:

> Thank you for notifying us, we have informed the support team and will follow up with their update accordingly. 
> 
> I do want to mention that if I recall correctly the activity: 123 Time Billing 0430 - "Minutes Used" is calculated using the following values. However, we will update you with the response from the support team after they are able to research further. 
> 
> (PNBLSInboundTimeUnits + PNBLSOutboundTimeUnits + PNBLSConfSetupUnits + PNBLSVMTimeUnits + PNBLSWrapTimeUnits) / 10

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 20, 2018, 08:49

Kimberly wrote:

>  I am following up on this ticket to see if you have any feedback?

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Mar 21, 2018, 02:25

@kdjstudios 

> I tried to look into the account's Human Readable but all of the calculated values show Zero, however it is showing usage on those calculated fields in the invoice..?

Hr report was fixed. So I guess she can get required info from that report.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 21, 2018, 07:55

Kimberly wrote:

> Thank you for your email.   
> 
> I looked at the transaction detail report and I am still not coming up with the correct amount.  I used what you had mentioned: 
> 
> (PNBLSInboundTimeUnits + PNBLSOutboundTimeUnits + PNBLSConfSetupUnits 
+ PNBLSVMTimeUnits + PNBLSWrapTimeUnits) / 10
> 
> I have attached the transaction detail report.  I highlighted the sections you mentioned above to add and then multiply by 10. 
> 
> Inbound Time/Units = 6,049 + Outbound Time/Units = 1,796 + VM Time/Units = 0 + Set/Qty/Time/Units = 45 + Agnt Conf Wrap Time/Units = 1,950  Total = 9,840 / 10 = 984.  
> 
> The invoice total units was 1009 it is still off by 25 units. 
> 
> Are you merging in any fax data? 
> Are you using a percent posting?  
> 
> Once you have the answer, please let me know so I can let my client know. 

[transdetail_lynnhousing_122717_012618.pdf](/uploads/d2be1dba7e79593f9956bd6d5c7ab4cd/transdetail_lynnhousing_122717_012618.pdf)

[invoice_123-67311_Febinvoice.pdf](/uploads/3a63d1366fcb92e581fe9176fb7684c6/invoice_123-67311_Febinvoice.pdf)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 21, 2018, 08:02

@smasih 

The HR report must still not be working. as its total for 400.82.

![image](/uploads/a7b070f78d919e0d2b1ef35ba25d30a1/image.png)

In addition to that the manual totals calculated do not equal the total on the invoice. Please resolve this.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Mar 22, 2018, 06:55

@kdjstudios 

System calculation is working fine as 400.82 came after subtracting allowance and multiplying excess rate. But the issue here is how it got different no. of units on the invoice. 

I don't know how system came up with those units and the data currently we have in system is different  So, please we need their **Usage file from december to jan** unformatted.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 22, 2018, 09:37

@smasih @tim.traylor @pchaudhary @rchauhan 

Please read the above posts. This is becoming quite persistent that we are having to continually repeat everything being provided. 

You may use the attached transdetail report in the above post to confirm usage. 

@tim.traylor Also I could have sworn that the uploaded usage files are on the server, Correct?

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Mar 23, 2018, 01:26

@kdjstudios 

We need a usage file, and yes it's uploaded to the server but that gets formatted. If the file was a .txt format it gets converted to .csv with some additional details added on to it. 

So we need that unformatted, please in order to narrow down the issue, which can be with:    
* Posting script
* Invoice calculation for transactions 
* Usage File

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 23, 2018, 08:24

@smasih 

I am confused.

- Why can you not use the already available usage file to check SAB calculations manually?
- The already uploaded usage file has all the information you need to complete this. 
 - If needed, remove the additional formatting. You have access to the posting script to see what type of file it takes in. This has now delayed progress on this ticket for over three days now.
- Why are we not saving a "original" copy of usage file that gets uploaded? This makes no sense?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 23, 2018, 08:26

Wrote to Kimberly:

> The support team can not seem to find the cause of this. They are requesting that you provide the original usage file that was uploaded to SA Billing. This will allow them to verify the original information that was submitted and also check the processes in SA Billing for any errors.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Mar 23, 2018, 09:54

assigned to @smasih

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 23, 2018, 10:39

Kimberly wrote:

> Here is the combined file that was uploaded to SAB.  
> 
> Please let me know if there is anything else you need. 

[combined_022818.txt](/uploads/b1f1c977fe9469bad228f9132de629a4/combined_022818.txt)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Mar 30, 2018, 15:41

@smasih 

Any updates?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 3, 2018, 16:09

@smasih 

Kimberly is requesting an update again.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 23, 2018, 08:22

@smasih @tim.traylor @pchaudhary @rchauhan 

It has been over a month and Kimberly has not received any update, she is reaching out for an update again. Please provide an update.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Apr 23, 2018, 08:50

Update sent to Kim:

> We do apologize as there is no update yet; there were other Higher Priority items that came up over the last few weeks. Once the support team is able to continue working on this we will inform you of an update.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 8, 2018, 15:59

@smasih @pchaudhary @rchauhan @tim.traylor @thomas.gentry 

Kimberly is asking for another update. Please discuss this in the morning meeting with @tim.traylor and post an update. Thank you.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on May 9, 2018, 24:21

@kdjstudios @tim.traylor 

prioritizing this ticket as this has been requested twice.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on May 9, 2018, 05:56

Hi [Kyle] ( @kdjstudios ),   

We have investigated the account properly & below are our findings.

* The result of the calculation according to the file provided by Kim is **402** which is not matching any of the results, neither Kim's nor system which leads that there may be some other file or there is a bug in the code. We request you to double check the file with Kim.

* The result of **402** comes after the following calculation - ``(PNBLSInboundTimeUnits + PNBLSOutboundTimeUnits + PNBLSConfSetupUnits + PNBLSVMTimeUnits + PNBLSWrapTimeUnits) / 10``  = ((2281 + 745 + 13 + 975).to_f / 10.0).ceil


* In the server we found this file ([012918.csv.20180201100204.csv](/uploads/2342ac71275b8505a8182728107ae753/012918.csv.20180201100204.csv)) uploaded in February, we are not able to find the account number with the resource ID **564** in this file.

* The posting script is using the same formula provided in the HR report. So, this concludes either there is a bug in the code which is very unlikely or posted usage file had different values.

**Can you please double check with Kim that which file was posted**

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 9, 2018, 08:47

I wrote:

> The support team was able to research this and found the following:
> 
> 1. The result of the calculation according to the file provided by Kim is 402 which is not matching any of the results. The support team requests you to double check the usage file that was uploaded SA Billing.
> 2. The result of 402 comes after the following calculation - (PNBLSInboundTimeUnits + PNBLSOutboundTimeUnits + PNBLSConfSetupUnits + PNBLSVMTimeUnits + PNBLSWrapTimeUnits) / 10  = ((2281 + 745 + 13 + 975).to_f / 10.0).ceil
> 3. In the server the support team found the attached file uploaded for February usage. However, The support team was not able to find the account number with the resource ID 564 in this file.
> 4. The upload script is using the same formula provided in the HR report. So, this concludes that most likely the posted usage file had different values.
> 
> Please let us know if you have any questions.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 9, 2018, 10:21

Kimberly Wrote:

> The csv file you sent me does not have my clients listed on it.  I suspect this is the wrong file.  Can you recheck your findings with the correct csv file and let me know? 

I wrote:

> I think you misunderstood. That is what we are asking of you in the # 1. To check which file was actually uploaded to SA Billing and please submitted it to us.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 9, 2018, 10:40

Kimberly wrote:

> I just went back to find the file that I sent on March 23 and it was the combined file uploaded to SAB and here it is. 
> 
> I am not sure where that csv file came from. (Support Team's CSV)

[combined_022818__1_.txt](/uploads/364f076b4191930a104fa3772ff0e3c3/combined_022818__1_.txt)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 9, 2018, 10:42

@smasih @pchaudhary @tim.traylor @thomas.gentry 

Please see the updated conversation between Kim and I. It would appear that the file the support team provided us is not for the correct site?

**Comment by kdjstudios** [comment]
In GitLab by @smasih on May 10, 2018, 07:24

@kdjstudios 

Thanks for posting the file Kyle, we got confused and pulled the report for `053` (mentioned at top) while we should have been looking for `123`.

 
we spent 2hrs on this issue but didn't reach to any conclusion. We are going to spend some more time on it  in figuring out the issue.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 10, 2018, 12:07

changed the description

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 10, 2018, 12:14

@smasih @pchaudhary @rchauhan @thomas.gentry 

Thank you for pointing out the typo in the main ticket. I have corrected this. However, I am slightly confused by how the rest of the comments, invoices, files, etc all mentioned 123 multiple times throughout the conversation, why did we choose to ignore those and look there?

Marking this for discussion. Please review this with @tim.traylor in the morning meetings. As we will want his input on how much time to spend on this.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on May 22, 2018, 07:07

@kdjstudios @tim.traylor 

Here is the usage file used for calculation of this invoice [invoice_123-67311__1_.pdf](/uploads/4630c0489d83a31a1a8bc28fa781d8df/invoice_123-67311__1_.pdf)     
usage file: [123_jan.txt](/uploads/012a755f022e8933a43052ab697f064d/123_jan.txt)

after investigating we found that the bug is in posting script.

here is the formula used to calculate line item **Minutes Used**  is `PNBLSInboundTimeUnits + PNBLSOutboundTimeUnits + PNBLSConfSetupUnits + PNBLSVMTimeUnits + PNBLSWrapTimeUnits`

|                | PNBLSInboundTimeUnits  |  PNBLSOutboundTimeUnits  |  PNBLSConfSetupUnits  |  PNBLSVMTimeUnits  |  PNBLSWrapTimeUnits | Total |
|----------------|------------------------|--------------------------|-----------------------|--------------------|---------------------|-------|
| Should be      | 62151                  | 1829                     | 45                    | 0                  | 1999                | 66024 |
| on calculation | 6215                   | 1829                     | 45                    | 0                  | 1999                | 10088 |


in `PNBLSInboundTimeUnits` the value should be 62151 but system is taking it as 6215.

The total is further divided by 10 (total/10)

We will continue to further debug the issue.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 22, 2018, 08:37

I wrote:

> We just want to send an update out that the support team is making progress in researching the issue and developing a solution. We will continue to update as the support team makes progress.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on May 24, 2018, 06:31

@kdjstudios @tim.traylor 

#### our previous findings were not accurate.

Sorry for the confusion, but the script is working fine. We checked in the usage file, the values imported were correct, **so the amount that was calculated in reported invoice is correct.**

However, the bug is how system is storing the info, which got us confused at first place. This is the same place where HR report is getting it's data from.

[Screenshot_from_2018-05-24_17-06-30](/uploads/ad306e69be9963ced69eb27371529b9b/Screenshot_from_2018-05-24_17-06-30.png)

In short, for users, script is working just fine for the invoice calculation but not working properly while updating the transaction data, which is primarily used for Human Readable report. 

**This issue is only spotted with Billerica (pinnacle switch)**

**We haven't tested this for Towne yet. We will let you know once checked.**

Thanks.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 24, 2018, 13:18

@pchaudhary @tim.traylor @thomas.gentry 

I believe in #760 it was mentioned that the HR report should not be using transaction/invoice data. It also looks like that has not yet been discussed with @tim.traylor as mentioned.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on May 25, 2018, 01:05

@kdjstudios 

> I believe in #760 it was mentioned that the HR report should not be using transaction/invoice data. It also looks like that has not yet been discussed with @tim.traylor as mentioned.

allow me to explain:

Hr report has two parts:

* Top section one is showing data in usage file.
  * This section is pulling data from the usage file which was saved in the system during the upload.
  * In @pchaudhary 's comment we are trying to highlight this section.
  * Pinnacle script for Billerica is working when it comes to transaction calculation but has issues pushing data symmetrically into the database (not sure if its posting script issue).
  * We have to research further whether it's posing script or existing sab code that is not compatible with the usage file format.

* Bottom section which is pulling invoice data   
  * What you suggested that has been already discussed with @tim.traylor and he agreed to the changes you suggested.     
  * But, we have not implemented that yet.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on May 25, 2018, 07:46

@tim.traylor 



> We have to research further whether it's posing script or existing sab code that is not compatible with the usage file format.

During the upload system creates a csv file which is then converted into hash(key value) and saved into db.
So the formatted file that is being generated could have different values. 

**But this does not affect:** 

* the invoice usage calculation
* Profitability Report
* Exception report 
* RevenuePerMinute

But does affect

* hr report

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 11, 2018, 09:20

@smasih @pchaudhary @tim.traylor 

Kim is requesting an update again. Please bring this up during discussion on Tuesday with @tim.traylor 

It would appear there has been progress made on the researching into finding the cause of this, but I do not see any proposed solution yet?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 11, 2018, 09:24

Updating priority. As this is a client facing issue and it has been over two months since reported.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Jun 12, 2018, 02:00

@kdjstudios @thomas.gentry @tim.traylor 

for today's discussion here is a brief explanation on where we left this ticket last

* There is a bug in parsing data from csv to database acceptable format (bson). We have only seen this happening with this site, maybe usage csv we are supplying is not properly formatted.

* **Client or billing/ Invoice is not affected by this.** Only HR report can be affected by this bug or usage format.

* The calculation [reported invoice](http://gitlab.aavaz.biz/AnswerNet/SABilling/uploads/d19d687e8317e10fa75a2c52a5764ce1/invoice_123-67311_Feb.pdf) has is absolutely correct.

* here is the [usage of jan](http://gitlab.aavaz.biz/AnswerNet/SABilling/uploads/012a755f022e8933a43052ab697f064d/123_jan.txt) to confirm it.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 12, 2018, 09:44

The Aavaz team will be researching and updating the code to correctly parse the data. @tim.traylor states you have access to all of the usage files needed to confirm and test this.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 18, 2018, 08:27

@smasih @pchaudhary @tim.traylor 

What is the latest update on this ticket?

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Jul 19, 2018, 05:29

@kdjstudios 

there are no updates on this at the moment.

root cause for this issue is not fixed **but the initial question of Kimberly was answered**

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 26, 2018, 10:41

@smasih @pchaudhary @tim.traylor 

Kim wrote:

> I am looking for an update on this ticket.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 30, 2018, 10:44

@smasih @pchaudhary @abakshi @rchauhan 

This was discussed during our internal tech meeting as Kim is escalating the ticket. Please review this with @tim.traylor in the morning meetings.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Jul 31, 2018, 07:25

@abakshi @kdjstudios 

Kim's query was answered in the comment http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/757#note_22018

There is data inconsistency compared to system data vs usage file since beginning, which we suspect is due to file format.

Here are the two usage files for the same switch (_pinnacle_). In case of Billerica if we scroll down we can find some data which seems like is not the part of usage file. 

[Towne_pinnacle_usage.csv](/uploads/9c98b4a54a258b30c838646a23e0f3ef/twn.csv)

VS

[Billerica (MA inbound) Usage.csv](/uploads/ea561dfbe71b9a9706cad5f12e62df81/bill_123.csv)

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 31, 2018, 08:55

@smasih @pchaudhary @tim.traylor @rchauhan @abakshi 

Thank you for the update, however the comment right after the one you list appears to clearly highlight that this was discussed and we are still awaiting your fix. What is the update on the resolution?

http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/757#note_22047 "The Aavaz team will be researching and updating the code to correctly parse the data."

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 31, 2018, 10:00

After reviewing with @tim.traylor during today's meeting we were able to determine that the export data has mixed data, it appears to have multiple reports. This is most likely an issue with their switch while running that report.  This data was not effecting the import process, but was effecting the Human Readable report. Tim suggests since thier site is converting shortly that we not do anything until then.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jul 31, 2018, 10:03

I wrote:

> The support team after reviewing the uploaded resource file provided earlier determined that the export data has mixed data, from what would appear to have multiple reports. This is most likely an issue with the switch while running that report.  This data was not effecting the import process, but was effecting the Human Readable report. Tim suggests since your site is converting shortly that we not do anything until then, as once switched over the VCC reports will work correctly.

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Aug 1, 2018, 24:44

removed milestone

**Comment by kdjstudios** [comment]
In GitLab by @smasih on Aug 1, 2018, 24:44

@kdjstudios 

can we replace the "urgent" tag with "paused" tag..?

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Aug 1, 2018, 15:14

@smasih 

Yes, lets do that. Marking in pause until site converts to VCC.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 14, 2018, 14:06

Site has not converted to VCC, but we have not heard any more complaints from the client. Marking for close.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 14, 2018, 14:49

closed
